File: codes/contest/leetcode/longest-palindromic-subsequence.cpp
/* coding:utf-8
 * Copyright (C) dirlt
 */

#include <string>
#include <utility>
using namespace std;
class Solution {
   public:
    int longestPalindromeSubseq(string s) {
        int n = s.size();
        int** dp = new int*[n];
        for (int i = 0; i < n; i++) {
            dp[i] = new int[n];
        }
        for (int i = 0; i < n; i++) {
            dp[i][i] = 1;
        }

        for (int k = 2; k <= n; k++) {
            for (int i = 0; i <= (n - k); i++) {
                int j = i + k - 1;
                if (s[i] == s[j]) {
                    if ((i + 1) <= (j - 1)) {
                        dp[i][j] = dp[i + 1][j - 1] + 2;
                    } else {
                        dp[i][j] = 2;
                    }
                } else {
                    dp[i][j] = max(dp[i + 1][j], dp[i][j - 1]);
                }
            }
        }
        int res = dp[0][n - 1];
        for (int i = 0; i < n; i++) {
            delete[] dp[i];
        }
        delete[] dp;
        return res;
    }
};

File: codes/contest/leetcode/longest-palindromic-subsequence.py
#!/usr/bin/env python
# coding:utf-8
# Copyright (C) dirlt

class Solution:
    def longestPalindromeSubseq(self, s):
        """
        :type s: str
        :rtype: int
        """

        n = len(s)
        if n == 0:
            return 0

        dp = []
        for i in range(n):
            dp.append([0] * n)
        for i in range(0, n):
            dp[i][i] = 1
        for k in range(2, n + 1):
            for i in range(0, n - k + 1):
                j = i + k - 1
                if s[i] == s[j]:
                    if (i + 1) <= (j - 1):
                        dp[i][j] = dp[i + 1][j - 1] + 2
                    else:
                        dp[i][j] = 2
                else:
                    dp[i][j] = max(dp[i + 1][j], dp[i][j - 1])
        return dp[0][n - 1]


if __name__ == '__main__':
    s = Solution()
    print((s.longestPalindromeSubseq('bbbab')))
    print((s.longestPalindromeSubseq('cbbd')))
    print((s.longestPalindromeSubseq(
        "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb")))
